### PR TITLE
Feature/rm 9251 remove net framework from re motion codebase

### PR DESCRIPTION
**Pull request by Frazerus**
https://re-motion.atlassian.net/browse/RM-9251

### CODE REVIEW

File: Remotion/Security/MSBuild.Tasks/Security.MSBuild.Tasks.csproj [hunk 1]
```diff
@@ -5,7 +5,7 @@
   </PropertyGroup>
```
Comment (MichaelKetting) [inline Remotion/Security/MSBuild.Tasks/Security.MSBuild.Tasks.csproj:5]: this project gets deleted.

File: Build/Build.cs [hunk 2]
```diff
@@ -149,14 +139,12 @@ public override void ConfigureProjects (ProjectsBuilder projects)
...
     projects.AddReleaseProject("Data.DomainObjects.Security");
     projects.AddReleaseProject("Data.DomainObjects.UberProfIntegration");
     projects.AddReleaseProject("Data.DomainObjects.Validation");
     projects.AddUnitTestProject("Data.DomainObjects.UnitTests", databaseTestConfiguration);
     projects.AddUnitTestProject("Data.DomainObjects.ObjectBinding.UnitTests", normalTestConfiguration);
     projects.AddUnitTestProject("Data.DomainObjects.ObjectBinding.IntegrationTests", databaseTestConfiguration);
-    projects.AddUnitTestProject("Data.DomainObjects.RdbmsTools.UnitTests", normalTestConfiguration);
     projects.AddUnitTestProject("Data.DomainObjects.Security.UnitTests", normalTestConfiguration);
     projects.AddUnitTestProject("Data.DomainObjects.UberProfIntegration.UnitTests", databaseTestConfiguration);
     projects.AddUnitTestProject("Data.DomainObjects.Web.IntegrationTests", databaseTestConfiguration);
```
Comment (MichaelKetting) [inline Build/Build.cs:150]: delete Security.MSBuild.Tasks entry

File: Remotion/Data/DomainObjects.Web.Test/Queries1.xml [hunk 3]
```diff
@@ -17,7 +17,7 @@
  ! along with re-motion; if not, see http://www.gnu.org/licenses.
 -->
 <queries
-    xmlns="http://www.re-motion.org/Data/DomainObjects/Queries/2.0" 
+    xmlns="http://www.re-motion.org/Data/DomainObjects/Queries/2.0"
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.Web.Test/Queries1.xml:20]: strictly speaking, this change is irrelevant. it's also helpful and there's enough noise that it won't matter.

File: Remotion/Data/DomainObjects.Web.Test/Queries2.xml [hunk 4]
```diff
@@ -17,7 +17,7 @@
  ! along with re-motion; if not, see http://www.gnu.org/licenses.
 -->
 <queries
-    xmlns="http://www.re-motion.org/Data/DomainObjects/Queries/2.0" 
+    xmlns="http://www.re-motion.org/Data/DomainObjects/Queries/2.0"
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.Web.Test/Queries2.xml:20]: strictly speaking, this change is irrelevant. it's also helpful and there's enough noise that it won't matter.

File: Remotion/Core/Reflection.UnitTests/CodeGeneration/MethodWrapperEmitterTests/SetUpFixture.cs [hunk 5]
```diff
@@ -65,17 +58,6 @@ public virtual void OneTimeSetUp ()
...
       Console.WriteLine("Tearing down MethodWrapperEmitterTests");
-#if NETFRAMEWORK && ENABLE_PEVERIFY
-
-      s_assemblyBuilder.Save(s_moduleBuilder.ScopeName);
-      var path = s_moduleBuilder.FullyQualifiedName;
-
-      s_assemblyBuilder = null;
-      s_moduleBuilder = null;
-
-      PEVerifier.CreateDefault().VerifyPEFile(path);
```
Comment (MichaelKetting) [inline Remotion/Core/Reflection.UnitTests/CodeGeneration/MethodWrapperEmitterTests/SetUpFixture.cs:76]: please check PEVerify
please check occurances of ENABLE_PEVERIFY

File: Remotion/Development/Web/UnitTesting/ExecutionEngine/WxeFunctionThreadAbortHelper.cs [hunk 6]
```diff
@@ -7,19 +7,12 @@ public static class WxeThreadAbortHelper
...
     {
-#if NETFRAMEWORK
-      Thread.CurrentThread.Abort();
-#else
       var exception = (ThreadAbortException)Activator.CreateInstance(typeof(ThreadAbortException), nonPublic: true)!;
       throw exception;
-#endif
     }
 
     public static void ResetAbort ()
```
Comment (MichaelKetting) [inline Remotion/Development/Web/UnitTesting/ExecutionEngine/WxeFunctionThreadAbortHelper.cs:14]: delete please. should be a separate commit since the tests get touched by this and we should verify the correctness. we could remove in a separate task/PR, if you want.

File: Remotion/Mixins/UnitTests/Core/CodeGeneration/AdHocCodeGenerator.cs [hunk 7]
```diff
@@ -139,9 +135,6 @@ public string Save ()
         }
       }
 
-#if NETFRAMEWORK && ENABLE_PEVERIFY
```
Comment (MichaelKetting) [inline Remotion/Mixins/UnitTests/Core/CodeGeneration/AdHocCodeGenerator.cs:142]: please leave a TODO comment
`TODO RM-9285: save generated assembly for use with IL verification.`

File: Remotion/Web/Development.WebTesting.UnitTests/HostingStrategies/Configuration/TestSiteLayoutConfigurationTest.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting.UnitTests/HostingStrategies/Configuration/TestSiteLayoutConfigurationTest.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/Configuration/IConfiguration/WebTestTestSiteLayoutDto.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/Configuration/IConfiguration/WebTestTestSiteLayoutDto.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/Configuration/IWebTestTestSiteLayoutSettings.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/Configuration/IWebTestTestSiteLayoutSettings.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/Configuration/Legacy/TestSiteLayoutConfigurationElement.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/Configuration/Legacy/TestSiteLayoutConfigurationElement.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/AspNetCoreHostingProcessWrapper.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/AspNetCoreHostingProcessWrapper.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/AspNetCoreHostingStrategy.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/AspNetCoreHostingStrategy.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/AspNetDockerContainerWrapper.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/AspNetDockerContainerWrapper.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/Configuration/HostingConfiguration.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/Configuration/HostingConfiguration.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/Configuration/ITestSiteLayoutConfiguration.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/Configuration/ITestSiteLayoutConfiguration.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/Configuration/TestSiteLayoutConfiguration.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/Configuration/TestSiteLayoutConfiguration.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/DockerContainerWrapperBase.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/DockerContainerWrapperBase.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/AspNetDockerContainerConfigurationParameters.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/AspNetDockerContainerConfigurationParameters.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/DockerCommandLineClient.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/DockerCommandLineClient.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/DockerContainerConfigurationParameters.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/DockerContainerConfigurationParameters.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/IDockerClient.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/IDockerClient.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/IisDockerContainerConfigurationParameters.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/IisDockerContainerConfigurationParameters.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/DockerHostingStrategy.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/DockerHostingStrategy.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/IisDockerContainerWrapper.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/IisDockerContainerWrapper.cs:1]: once we're done with RM-9277 this change should fall away

File: Remotion/Web/Development.WebTesting/HostingStrategies/ProcessShutdownUtility.cs
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/ProcessShutdownUtility.cs:1]: once we're done with RM-9277 this change should fall away

File: SharedSource/Core/Utilities/EqualityUtility.cs [hunk 8]
```diff
@@ -179,7 +179,7 @@ public static int GetXorHashCode (IEnumerable objects)
     /// Similar to <see cref="Equals{T}"/>, but without any boxing (better performance). 
     /// Equatable objects implement the <see cref="IEquatable{T}"/> interface. 
     /// </remarks>
-    public static bool EqualsEquatable<T> ([AllowNull] T a, [AllowNull] T b)
+    public static bool EqualsEquatable<T> (T? a, T? b)
```
Comment (MichaelKetting) [inline SharedSource/Core/Utilities/EqualityUtility.cs:182]: revert please. AllowNullAttribute is part of nullable refernece type semantics

File: SharedSource/Core/Utilities/EqualityUtility.cs [hunk 9]
```diff
@@ -227,7 +227,7 @@ public static bool NotNullAndSameType<T> (T a, [NotNullWhen(true)] T? b)
     /// <remarks>
     /// Similar to <see cref="object.Equals(object,object)"/>, only with less boxing going on (better performance).
     /// </remarks>
-    public static bool Equals<T> ([AllowNull] T a, [AllowNull] T b)
+    public static bool Equals<T> (T? a, T? b)
```
Comment (MichaelKetting) [inline SharedSource/Core/Utilities/EqualityUtility.cs:230]: revert please. AllowNullAttribute is part of nullable refernece type semantics

File: SharedSource/UnitTests/Development/Core/UnitTesting/PEVerifierTest.cs
Comment (MichaelKetting) [inline SharedSource/UnitTests/Development/Core/UnitTesting/PEVerifierTest.cs:1]: PEVerifier and PEVerifierTest gets dropped completely.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
Code is now reviewed except for the actual web applications.

File: Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs [hunk 10]
```diff
@@ -3,9 +3,7 @@
 using System;
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs:3]: there's one `NETFRAMEWORK` left in line 50

File: Remotion/Core/Reflection.UnitTests/CodeGeneration/MethodWrapperEmitterTests/SetUpFixture.cs (hunk 5 above)
Comment (Frazerus) [reply Remotion/Core/Reflection.UnitTests/CodeGeneration/MethodWrapperEmitterTests/SetUpFixture.cs:76]: Do we want to have no occurrences of this? What is the goal with checking it?

Comment (Frazerus) [reply Remotion/Core/Reflection.UnitTests/CodeGeneration/MethodWrapperEmitterTests/SetUpFixture.cs:76]: Theres a comment about this lower down, I guess we want to remove all usages of it.

File: SharedSource/Core/Utilities/EqualityUtility.cs (hunk 8 above)
Comment (Frazerus) [reply SharedSource/Core/Utilities/EqualityUtility.cs:182]: I really dont like this.

Comment (Frazerus) [reply SharedSource/Core/Utilities/EqualityUtility.cs:182]: I think there was more of a reason to change this than simple preference, I think some builds failed? Doesnt happen anymore though, so its alright.

File: Remotion/Development/Web/UnitTesting/ExecutionEngine/WxeFunctionThreadAbortHelper.cs (hunk 6 above)
Comment (Frazerus) [reply Remotion/Development/Web/UnitTesting/ExecutionEngine/WxeFunctionThreadAbortHelper.cs:14]: The whole file or just the reset abort?

File: Remotion/Data/DomainObjects.PerformanceTests/Root.cs [hunk 11]
```diff
@@ -33,42 +33,6 @@ private Root ()
     {
     }
 
-#if NETFRAMEWORK
```
[the remark this answers is not in the document]
Comment (Frazerus) [reply Remotion/Data/DomainObjects.PerformanceTests/Root.cs:36]: I think the reason is that the TestSdk generates a Program.cs with a separate Main function, meaning we cannot build this in .NET

[the remark this answers is not in the document]
Comment (Frazerus) [reply Remotion/Data/DomainObjects.PerformanceTests/Root.cs:36]: I'm not sure how to fix this though.

File: Directory.Packages.props [hunk 12]
```diff
@@ -16,6 +14,8 @@
     <PackageVersion Include="Fork.Coypu" Version="3.4.0" />
     <PackageVersion Include="JetBrains.Annotations" Version="2023.3.0" />
     <PackageVersion Include="log4net" Version="2.0.17" />
+    <PackageVersion Include="Microsoft.Build.Framework" Version="17.8.3" />
+    <PackageVersion Include="Microsoft.Build.Utilities.Core" Version="17.8.3" />
```
[the remark this answers is not in the document]
Comment (Frazerus) [reply Directory.Packages.props:18]: I'm not sure, I thought this fixed some errors during the changes, but removing it now changed nothing.

File: Remotion/Development/Web/UnitTesting/ExecutionEngine/WxeFunctionThreadAbortHelper.cs (hunk 6 above)
Comment (MichaelKetting) [reply Remotion/Development/Web/UnitTesting/ExecutionEngine/WxeFunctionThreadAbortHelper.cs:14]: the RestAbort method since it's blank

File: Remotion/Data/DomainObjects.PerformanceTests/Root.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Data/DomainObjects.PerformanceTests/Root.cs:36]: aaaaaah... that's the reason. okay. when please keep the file but comment out the entire main method.

File: Directory.Packages.props (hunk 12 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Directory.Packages.props]: I found the reason later on but forgot to track back to here: it was added after switching Security.MSBUild.Tasks project from Framework to .NET 8. Since we have now deleted that project, it's all good and we don't need the lines any more.

File: SecurityManager/Clients.Web.Test/Program.cs
Comment (MichaelKetting) [inline SecurityManager/Clients.Web.Test/Program.cs:1]: Reminder: merge RM-9758 first.

File: Remotion/ObjectBinding/Web.Test/Program.cs
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.Test/Program.cs:1]: Reminder: merge RM-7954 first

File: Remotion/Data/DomainObjects.Web.Test/Program.cs
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.Web.Test/Program.cs:1]: Reminder: merge RM-7956 first.

**Review by MichaelKetting — APPROVED** [review]
okay, basically we want to first do rm-9277 and then update the test matrix to only run .net 8 tests in a dedicated PR.
then we switch the web applications to .net 8 in dedicated PR.
finally, this PR.

File: SharedSource/Core/Utilities/EqualityUtility.cs (hunk 8 above)
Comment (Frazerus) [reply SharedSource/Core/Utilities/EqualityUtility.cs]: I'm not sure whether I couldn't find a case where it matters or that it has been fixed in the meantime. This is with `T?`:

image

Comment (MichaelKetting) [reply SharedSource/Core/Utilities/EqualityUtility.cs]: > I'm not sure whether I couldn't find a case where it matters or that it has been fixed in the meantime. 

Hmm... I think you're correct about this being a c# version thing: https://endjin.com/blog/2022/02/csharp-10-generics-nullable-references-improvements-allownull

For tracking purposes, I think a separate task and PR would be better. While this is a very small change on the surface, when reading the semantics, you only understand them with checking up on the change in the c# compiler and a deep dive into why it is okay. Please extract into separate PR.

File: Remotion/Core/Core/Core.Core.csproj [hunk 13]
```diff
@@ -15,21 +15,15 @@
...
     <PackageReference Include="Remotion.ThirdPartyDependency.CommonServiceLocator" />
-  </ItemGroup>
-  <ItemGroup Condition="'$(IsDotNetFramework)' == 'true'">
-    <Reference Include="System.Configuration" />
-  </ItemGroup>
-  <!-- Including System.Configuration.ConfigurationManager is required because log4net takes a hard dependency on an older version of this assembly. -->
-  <ItemGroup Condition="'$(IsDotNetFramework)' == 'false'">
     <PackageReference Include="System.Configuration.ConfigurationManager" />
   </ItemGroup>
+  <!-- Including System.Configuration.ConfigurationManager is required because log4net takes a hard dependency on an older version of this assembly. -->
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Core.Core.csproj:26]: This comment section needs to be prepared to match original structure

File: Remotion/Core/UnitTests/Core.UnitTests.csproj [hunk 14]
```diff
@@ -23,36 +23,32 @@
...
   -->
-  <ItemGroup Condition="'$(IsDotNetFramework)' == 'false'">
-    <PackageReference Include="System.CodeDom" GeneratePathProperty="true" />
-    <PackageReference Include="System.Configuration.ConfigurationManager" GeneratePathProperty="true" />
-  </ItemGroup>
   <Target Name="CopySystemConfigurationConfigurationManager" AfterTargets="ResolveAssemblyReferences" Condition="'$(PkgSystem_Configuration_ConfigurationManager)' != ''">
-    <ItemGroup Condition="'$(IsDotNetFramework)' == 'false'">
+    <ItemGroup>
       <ReferenceCopyLocalPaths Include="$(PkgSystem_Configuration_ConfigurationManager)/lib/net8.0/System.Configuration.ConfigurationManager.dll"/>
     </ItemGroup>
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Core.UnitTests.csproj:48]: This section needs to be restored to it's original ordering to preserve the comments set for the respective parts of the dependency list

File: Remotion/Mixins/MixerTools/Mixins.MixerTools.csproj [hunk 15]
```diff
@@ -12,6 +12,8 @@
     <PackageId>Remotion.Mixins.MixerTools</PackageId>
     <RootNamespace>$(PackageId)</RootNamespace>
     <Nullable>disable</Nullable>
+    <!-- NuGet Package for MixerTools is disabled until APIs are fixed -->
+    <IsPackable>false</IsPackable>
```
Comment (MichaelKetting) [inline Remotion/Mixins/MixerTools/Mixins.MixerTools.csproj:16]: Check nightly build package output

File: Remotion/Data/DomainObjects.RdbmsTools/Data.DomainObjects.RdbmsTools.csproj [hunk 16]
```diff
@@ -11,6 +11,8 @@
     <PackageId>Remotion.Data.DomainObjects.RdbmsTools</PackageId>
     <RootNamespace>$(PackageId)</RootNamespace>
     <Nullable>disable</Nullable>
+    <!-- NuGet Package for MixerTools is disabled until APIs are fixed -->
+    <IsPackable>false</IsPackable>
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.RdbmsTools/Data.DomainObjects.RdbmsTools.csproj:15]: Check nightly build package output

**Review by MichaelKetting — APPROVED** [review]
